Add UI scaling when using GPU with stretched mode.

dbolya opened the pull request:
This changes the previous behavior, which instead of just scaling the UI texture, rendered the entire game into a smaller framebuffer and stretched that. This is much closer to actual UI scaling, in that we render the world view at its native resolution and only scale up the UI.

This should support all the options in the stretched mode menu as well as both fixed and resizeable mode.

Before:
![image](https://user-images.githubusercontent.com/19721752/48924130-bde9e800-ee69-11e8-921a-3273fdae9e7a.png)

After:
![image](https://user-images.githubusercontent.com/19721752/48924106-8c711c80-ee69-11e8-9a9b-d10923c17230.png)

devLotto commented on runelite-client/src/main/java/net/runelite/client/plugins/gpu/GpuPlugin.java:
```diff
@@ -879,10 +807,27 @@ else if (lastStretchedDimensions != null)
+				Dimension dim = client.getStretchedDimensions();
+				renderCanvasHeight = dim.height;
+
+				float scaleFactorY = dim.height / (float) canvasHeight;
```
Use dim.getHeight() instead of dim.height and you won't have to cast canvasHeight to float.

dbolya replied:
Good idea, changed.

devLotto commented on runelite-client/src/main/java/net/runelite/client/plugins/gpu/GpuPlugin.java:
```diff
@@ -991,7 +921,18 @@ private void drawUi(final int canvasHeight, final int canvasWidth)
+		{
+			Dimension dim = client.getStretchedDimensions();
+			renderCanvasWidth = dim.width;
+			renderCanvasHeight = dim.height;
```
Maybe just do
```java
if (client.isStretchedEnabled())
{
	Dimension dim = client.getStretchedDimensions();
	gl.glViewport(0, 0, dim.width, dim.height);
}
else
{
	gl.glViewport(0, 0, canvasWidth, canvasHeight);
}
```

dbolya replied:
Fixed.

devLotto commented on runelite-client/src/main/java/net/runelite/client/plugins/gpu/GpuPlugin.java:
```diff
@@ -1021,6 +962,14 @@ private void drawUi(final int canvasHeight, final int canvasWidth)
 		gl.glBindTexture(gl.GL_TEXTURE_2D, interfaceTexture);
 		gl.glUniform1i(uniTex, 0);
 
+		// If necessary, change the sampling function. Honestly, this should be using OpenGL
```
I guess "change the sampling function" is what the code does, and what follows in the comment is potential improvement. Maybe those two parts can be split up with a newline to make it less confusing what the comment is trying to convey.

What the code does can probably be clarified, maybe "Sets the sampling function used when stretching the UI".

dbolya replied:
Yeah comment there is a bit cryptic. Fixed that (and the line break). Also I was talking about [sampler objects](https://www.khronos.org/opengl/wiki/Sampler_Object). That's the _correct_ way to do it, but meh. Should I even leave that part of the comment in?

oabdellatif commented:
If this is merged, I think the name of the plugin should be changed to "Interface Scaling" to better reflect its new behavior. "Stretched Mode" could be misinterpreted to still refer to upscaling the contents of the game window.

devLotto commented:
@oabdellatif It still stretches the scene on CPU so the name should be kept.

Adam- commented:
I think the text looking weird is due to the stretching method since if you turn on increased perf mode it affects it.

dbolya commented:
It looks like it might not be possible to _perfectly_ position the viewport. For instance, if you had a scale factor of 1.5 and an x offset in fixed mode of 1 pixel, then the viewport would need to start at x=1.5. Since we can only give integer values to glViewport, I have it start at x=1.

Now the that means that objects in the UI are up to 1 pixel off of objects in the world view. I'm thinking that's an acceptable amount of error?

I also added some padding to get rid of the black borders in fixed mode, so that should be fixed:
![image](https://user-images.githubusercontent.com/19721752/48963443-26f35d80-ef47-11e8-9f3e-2aa5962ce481.png)

dbolya commented:
Pushed a change that should make that one pixel wide text more legible by always using nearest neighbors for the alpha channel.

Before:
![image](https://user-images.githubusercontent.com/19721752/48964022-7724ec80-ef54-11e8-92b3-5fa55763146f.png)

After:
![image](https://user-images.githubusercontent.com/19721752/48964024-7c823700-ef54-11e8-8245-c445aca7f931.png)

This is kind of an imperfect solution because all the text is still darker (since the pixels around the text are black with alpha 0 and opengl blends that black in anyway), but I guess it's better.

dbolya commented:
So I did some searching and it looks like the exact problem we were having is best solved by pre-multiplying the alpha into the color and then using `glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA)`.

So I implemented that and reverted the previous shader change. Looks better now:
![image](https://user-images.githubusercontent.com/19721752/48967381-20d79e00-ef94-11e8-8312-d6e4aa6d4112.png)

It's blurrier, but that's because it's properly doing bilinear sampling now (as in now it looks the same as stretched cpu mode).

Also I had to sort of unpack and repack the pixels array to achieve this. Let me know if there is a better way / if it's not a good idea to modify the pixels array directly.

outpoints commented:
This is exactly what I've wanted!

Good luck! If you need any help feel free to post some code and I'll take a look! :)

devLotto commented:
How is spending milliseconds iterating a pixel array better than just using an FBO?

dbolya commented:
It's not. Actually in the process of implementing a second pass with an FBO right now.

dbolya commented:
Ok I added back the FBO and moved the alpha pre-multiplication step to the GPU. I had to make this a new shader instead of using the old one and a boolean uniform because rendering to a framebuffer texture is a little different. I also only do the first pass when scaling is on and the scaling is linear.

Man, that was a lot of code for such a simple thing, but we do two passes now if we want to use the first pass for anything extra in the future.

devLotto commented:
Works well.

Adam- commented:
Thanks!
